### PR TITLE
Onion: multilayered concentric sphere and ellipsoid shapes

### LINKED ISSUE #293: Multilayered sphere/ellipsoid
The new shape `coated2`, corresponding to three concentric layers, will be added soon (#291). The natural extension of this shape is arbitrary number of concentric spheres (onion), with command line like:
```
-shape onion <d2/d> [<d3/d> ...]
```
where d=d1 is the diameter of the outer sphere, and d1,d2,... is the non-increasing sequence of diameters. The number of layers is limited only by `MAX_NMAT`, but this will be relaxed when #12 is solved.

For a reference - https://github.com/adda-team/adda/wiki/AddingShape

When this shape is implemented, `coated2` will become deprecated, while `coated` will remain, since it additionally allows for non-concentric shapes (shift of the core from the center).

One more step is to make concentric ellipsoid:
```
-shape onion_ell <y/x> <z/x> <d2/d> [<d3/d> ...]
```

And one more idea is confocal spheroid (not the same as concentric), which is sometimes used in methods based on separation of variables (but I am not sure that it is relevant for any practical problems).

### PR DESCRIPTION
**Pull request by jeromefung**
### Description
Add shapes `onion` and `onion_ell` for multilayered concentric spheres and ellipsoids, respectively.

`onion` extends existing shapes `coated` and `coated2` to allow for arbitrary number of concentric layers, not just 2 or 3. (The upper limit is set by `MAX_NMAT`.) Similarly, `onion_ell` extends existing `ellipsoid` shape to allow for an arbitrary number of concentric ellipsoidal layers. These new shapes might be useful for modeling particles whose refractive index varies continuously by approximating the variation stepwise.

### Related issues
Fixes #293. (Confocal spheroids, briefly mentioned in #293, are not implemented.)

### Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist

- [x] I have read the [contributing guidelines](https://github.com/adda-team/adda/wiki/InstructionCommitters)
- [x] The new code complies with the existing [code style](https://github.com/adda-team/adda/wiki/CodeStyleGuide).
- [x] The code compiles correctly in all relevant regimes (at least, `make seq`)
- [x] The new code does not rely on any Fortran or C++ sources or is disabled by `NO_FORTRAN` or `NO_CPP` preprocessor macros, respectively.
- [x] The change neither adds or removes files; otherwise, these changes are reflected in `README.md` in corresponding folders.  
- [x] No warnings appear during debug compilation (at least, `make seq OPTIONS=DEBUG`, but better `devtools/build_debug`) or they are discussed below
- [x] Tests pass locally with my changes (at least, `sh comp2exec seq` in `tests/2exec`, but better `devtools/test_new [seq]`). If any errors appear, they are discussed below.
- [x] I have added tests that prove my fix is effective or that my feature works. And these tests pass. This includes new command line in suite files in `tests/2exec` (and potentially new ignore patterns). In some cases, it is desirable to add new tests to `tests/equiv`. 
- [x] I have added/extended necessary documentation (if appropriate). If suggesting changes to the manual, I have used "Track changes" in the doc file.
- [x] I have looked through all changes introduced by this pull request (line by line), using `git diff` or, better, some GUI tool, to ensure that no unexpected changes are introduced.

* Some tests in `comp2exec seq` fail, but same failures occur in `master`.
* Descriptions of the new shapes have been added to the manual. However, I did not attempt to update Fig. 3 (pictures of predefined shapes) since it is already very crowded and a rearrangement would be needed to fit in additional shapes.

### CODE REVIEW

File: src/make_particle.c [hunk 1]
```diff
@@ -2211,6 +2271,38 @@ void MakeParticle(void)
...
+				// only consider dipoles inside the sphere				
+				if (r2<=0.25) {
+					// first test if inside the core, then test if in layers from the outside in
+					if (r2<=onion_r2[nlayers-2]) mat=nlayers-1; // inside core 
+					else {
+						for (layer_ctr=0;layer_ctr<(nlayers-1);layer_ctr++) {
+							if (r2>onion_r2[layer_ctr]) {
+								mat=layer_ctr;
+								break;
+							}
```
Comment (myurkin) [inline src/make_particle.c:2285]: Have you considered any acceleration for such search? 

Actually, I tried running it with 250 layers, and the time for make_particle was always at least 10 times smaller that a single iteration. So definitely not a bottleneck. However, the code may even be neater if bsearch function is used.

**Review by myurkin — COMMENTED** [review]
Jerome, as I said before, your contribution is really comprehensive. Thanks for that! I will have some minor changes here and there (mostly in help strings and such, I will commit them later), but here are couple of questions, which you may want to address.

**Comment by jeromefung** [comment]
Thanks very much -- I'll respond to your comments over the next day or two.

**Comment by myurkin** [comment]
I have added various improvement. Feel free to comment on them (or add anything). Apart of that, we only need to finalize `make_particle.c`

File: src/make_particle.c [hunk 2]
```diff
@@ -1687,6 +1689,64 @@ void InitShape(void)
...
 			Nmat_need=1;
 			break;
+		case SH_ONION: {
+			char *layer_str=NULL;
+			// determine number of layers by searching for nonzero values passed in to sh_pars
+			for (i=0;i<MAX_N_SH_PARMS;i++) {
+				if (sh_pars[i]==0) {
+					nlayers=i+1;
+					break;
+					}
```
[the remark this answers is not in the document]
Comment (jeromefung) [reply src/make_particle.c]: It was not my intention to allow any arguments to be 0 -- I was trying to determine the number of layers by counting the number of non-zro parameters in `sh_pars`. Looking again at this section of the code, I realized that there's a much simpler way: the variable `sh_Npars`. 

I've pushed a fix for this and will look at the bsearch issue by the end of the week.

File: src/make_particle.c (hunk 1 above)
Comment (jeromefung) [reply src/make_particle.c:2275]: Respectfully, I'm not sure that using `bsearch` would significantly improve performance (given that `make_particle` is not an overall bottleneck in running ADDA) or would make this easier to read. Unless I'm missing something, `bsearch` would find an exact match between the `r2` value of a given dipole and an element of `onion_r2`, but that's not what is needed here. Moreover, using `bsearch` would require writing a comparison function.

**Comment by jeromefung** [comment]
Your changes to files other than `make_particle.c` look good to me, and I think I've addressed your questions about `make_particle.c`. Please let me know if you have any other questions or concerns; I have no further issues that would preclude merging.

**Review by myurkin — APPROVED** [review]

**Comment by myurkin** [comment]
Jerome, I have added a few (cosmetic) improvements. Can you please quickly check that the latest version works on your side? And then I will merge.

**Pull request merged**